### PR TITLE
One sided relationship progress

### PR DESCRIPTION
**Pull request by nollymar**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactory.java [hunk 1]
```diff
@@ -77,6 +78,7 @@ void deleteByContent(Contentlet contentlet, Relationship relationship, List<Cont
 
     List<Relationship> dbAll(String orderBy, String contentTypeInode) throws DotHibernateException;
 
-
+    List<Relationship> getOneSidedRelationships(final ContentType contentType, int limit,
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactory.java]: add some doc

File: dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java [hunk 2]
```diff
@@ -552,6 +554,25 @@ public void addRelationship(String parent,String child, String relationType)thro
...
-    
 
+    @Override
+    public List<Relationship> getOneSidedRelationships(final ContentType contentType, int limit, final int offset) throws DotDataException {
+        DotConnect dc = new DotConnect();
+        StringBuilder sql = new StringBuilder();
+
+        if (limit == 0)
+            throw new DotDataException("limit param must be more than 0");
+        limit = (limit < 0) ? 10000 : limit;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java]: parameter should be read-only I would try to create a new var here

File: dotCMS/src/main/java/com/dotcms/contenttype/model/field/RelationshipField.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import com.dotcms.repackage.com.google.common.collect.ImmutableList;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import java.util.Collection;
+import java.util.List;
+import org.immutables.value.Value;
+
+/**
+ * @author nollymar
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/model/field/RelationshipField.java]: add some doc

File: dotCMS/src/main/java/com/dotcms/contenttype/transform/relationship/RelationshipTransformer.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.dotcms.contenttype.transform.relationship;
+
+import com.dotmarketing.portlets.structure.model.Relationship;
+import java.util.List;
+
+public interface RelationshipTransformer {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/transform/relationship/RelationshipTransformer.java]: add some doc

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    @JSONP
+    @NoCache
+    @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+    public final Response getOneSidedRelationships(
+            @QueryParam("contentTypeId") final String contentTypeId,
+            @QueryParam(PaginationUtil.PAGE) final int page,
+            @QueryParam(PaginationUtil.PER_PAGE) @DefaultValue("-1") final int perPage,
+            @Context final HttpServletRequest request) throws Throwable {
+        Logger.debug(this,
+                "Getting the possible relationships and content types");
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java]: I would include the contentTpeId in the log

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        final PaginationUtil paginationUtil = new PaginationUtil(new RelationshipPaginator(user));
+
+        try {
+
+            final Map<String, Object> params = map(RelationshipPaginator.CONTENT_TYPE_PARAM,
+                    contentType);
+            return paginationUtil.getPage(request, user, null, page, perPage, params);
+        } catch (PaginationException e) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java]: there is already something that handle most of the cases for exceptions
return ResponseUtil.mapExceptionResponse(e);

File: dotCMS/src/main/java/com/dotcms/util/pagination/RelationshipPaginator.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.portlets.structure.model.Relationship;
+import com.dotmarketing.util.PaginatedArrayList;
+import com.liferay.portal.model.User;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * @author nollymar
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/util/pagination/RelationshipPaginator.java]: Add some doc

File: dotCMS/src/main/java/com/dotmarketing/business/RelationshipAPI.java [hunk 8]
```diff
@@ -82,5 +83,6 @@ void deleteByContent(Contentlet contentlet, Relationship relationship, List<Cont
 
   void addRelationship(String parent, String child, String relationType) throws DotDataException;
 
-
+  List<Relationship> getOneSidedRelationships(final ContentType contentType, final int limit,
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/RelationshipAPI.java]: Add some doc

**Review by jgambarios — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java [hunk 9]
```diff
@@ -552,6 +554,25 @@ public void addRelationship(String parent,String child, String relationType)thro
...
     }
-    
 
+    @Override
+    public List<Relationship> getOneSidedRelationships(final ContentType contentType, int limit, final int offset) throws DotDataException {
+        DotConnect dc = new DotConnect();
+        StringBuilder sql = new StringBuilder();
+
+        if (limit == 0)
+            throw new DotDataException("limit param must be more than 0");
```
Comment (oarrietadotcms) [inline dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java]: Could we print more information?, for example ct name, etc.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        final ContentTypeAPI contentTypeAPI = APILocator.getContentTypeAPI(user);
+        ContentType contentType;
+
+        if (UtilMethods.isSet(contentTypeId)) {
+            //Validate if contentTypeId is valid
+            try {
+                contentType = contentTypeAPI.find(contentTypeId);
+
+            } catch (DotSecurityException | DotDataException e) {
+                return ExceptionMapperUtil
```
Comment (oarrietadotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java]: please log the error before returning the Exception.

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+                        .createResponse(map("message", "Invalid Content Type"),
+                                "Invalid Content Type",
+                                Status.BAD_REQUEST);
+            }
+        } else {
+            return ExceptionMapperUtil
```
Comment (oarrietadotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/relationships/RelationshipsResource.java]: please log the error before returning the Exception.

File: dotCMS/src/main/java/com/dotcms/util/pagination/RelationshipPaginator.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            final List<Relationship> relationships = relationshipAPI
+                    .getOneSidedRelationships((ContentType) params.get(CONTENT_TYPE_PARAM), limit, offset);
+
+            //Adding one sided relationships
+            result.add(relationships.stream()
+                    .collect(Collectors.toMap(Relationship::getInode, r -> r)));
+
+            return result;
+        } catch (DotDataException e) {
+            throw new DotRuntimeException(e);
```
Comment (oarrietadotcms) [inline dotCMS/src/main/java/com/dotcms/util/pagination/RelationshipPaginator.java]: please log the error before returning the Exception.

File: dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java (hunk 9 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java]: Maybe use DotPreconditions.checkArgument(limit>0). This throws IllegalArumentException

Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/contenttype/business/RelationshipFactoryImpl.java]: @oarrietadotcms creo que no hace falta ya que es solo validando el limit. El ct name sería irrelevante

**Pull request closed without merging**